### PR TITLE
Ports soft panic bunker from TG

### PR DESCRIPTION
**Pull request by ike709**
## About The Pull Request

Ports:
https://github.com/tgstation/tgstation/pull/44319 (as a bonus)
https://github.com/tgstation/tgstation/pull/54465
https://github.com/tgstation/tgstation/pull/61837

Crossed asked for this.

~~Needs testing and general code cleanup, I finished this 5 minutes before my shift ended.~~

~~I'll write the CL later.~~

## Changelog
:cl:
tweak: Improved the antag rolling prevention.
admin: Changed how the panic bunker works. Admins can now configure it to make first-time players interview to receive access instead of being blocked.
/:cl:

### CODE REVIEW

**Comment by ike709** [comment]
DNM pending testing.

**Comment by Crossedfall** [comment]
The interview system triggers as expected for the player but there is no tab/UI element for the admin to interact with the interviewee.

File: code/controllers/configuration/config_entry.dm [hunk 1]
```diff
@@ -118,6 +118,19 @@
...
+/datum/config_entry/str_list
+	abstract_type = /datum/config_entry/str_list
+	config_entry_value = list()
+	dupes_allowed = TRUE
+
+/datum/config_entry/str_list/ValidateAndSet(str_val)
+	if (!VASProcCallGuard(str_val))
+		return FALSE
+	str_val = trim(str_val)
+	if (str_val != "")
```
Comment (utilbird) [inline code/controllers/configuration/config_entry.dm]:
```suggestion
	if (length(str_val))
```

File: code/modules/interview/interview_manager.dm [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+  * * id - The ID of the interview to find
+  */
+/datum/interview_manager/proc/interview_by_id(id)
+	if (!id)
+		return
+	for (var/ckey in open_interviews)
+		var/datum/interview/I = open_interviews[ckey]
+		if (I?.id == id)
+			return I
+	for (var/datum/interview/I in closed_interviews)
```
Comment (utilbird) [inline code/modules/interview/interview_manager.dm]:
```suggestion
	for (var/datum/interview/I as() in closed_interviews)
```

File: code/modules/interview/interview_manager.dm [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+  */
+/datum/interview_manager/proc/release_from_cooldown(ckey)
+	cooldown_ckeys -= ckey
+
+/**
+  * Dequeues the first interview from the interview queue, and updates the queue positions of any relevant
+  * interviews that follow it.
+  */
+/datum/interview_manager/proc/dequeue()
+	if (interview_queue.len == 0)
```
Comment (utilbird) [inline code/modules/interview/interview_manager.dm]:
```suggestion
	if (!length(interview_queue))
```

File: code/modules/interview/interview_manager.dm [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+		return
+
+	// Get the first interview off the front of the queue
+	var/datum/interview/to_return = interview_queue[1]
+	interview_queue -= to_return
+
+	// Decrement any remaining interview queue positions
+	for(var/datum/interview/i in interview_queue)
+		i.pos_in_queue--
```
Comment (utilbird) [inline code/modules/interview/interview_manager.dm]:
```suggestion
	for(var/datum/interview/I as() in interview_queue)
		I.pos_in_queue--
```

File: code/modules/interview/interview_manager.dm [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+/datum/interview_manager/proc/dequeue_specific(datum/interview/to_dequeue)
+	if (!to_dequeue)
+		return
+
+	// Decrement all interviews in queue past the interview being removed
+	var/found = FALSE
+	for (var/datum/interview/i in interview_queue)
+		if (found)
+			i.pos_in_queue--
+		if (i == to_dequeue)
```
Comment (utilbird) [inline code/modules/interview/interview_manager.dm]:
```suggestion
	for (var/datum/interview/I as() in interview_queue)
		if (found)
			I.pos_in_queue--
		if (I == to_dequeue)
```

File: code/modules/mob/dead/new_player/new_player.dm [hunk 6]
```diff
@@ -476,3 +484,58 @@
...
+	for (var/v in client.verbs)
+		var/procpath/verb_path = v
+		//if (!(verb_path in GLOB.stat_panel_verbs))
+		remove_verb(client, verb_path)
+
+	// Then remove those on their mob as well
+	for (var/v in verbs)
+		var/procpath/verb_path = v
+		//if (!(verb_path in GLOB.stat_panel_verbs))
+		remove_verb(src, verb_path)
```
Comment (utilbird) [inline code/modules/mob/dead/new_player/new_player.dm]: Does this actually casting do anything? It appears to be arbitrary

File: code/modules/interview/interview_manager.dm [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+	to_queue.pos_in_queue = interview_queue.len + 1
+	interview_queue |= to_queue
+
+	// Notify admins
+	var/ckey = to_queue.owner_ckey
+	log_admin_private("Interview for [ckey] has been enqueued for review. Current position in queue: [to_queue.pos_in_queue]")
+	var/admins_present = send2tgs_adminless_only("panic-bunker-interview", "Interview for [ckey] enqueued for review. Current position in queue: [to_queue.pos_in_queue]")
+	if (admins_present <= 0 && to_queue.owner)
+		to_chat(to_queue.owner, "<span class='notice'>No active admins are online, your interview's submission was sent through TGS to admins who are available. This may use IRC or Discord.</span>")
+	for(var/client/X in GLOB.admins)
```
Comment (utilbird) [inline code/modules/interview/interview_manager.dm]:
```suggestion
	for(var/client/X as() in GLOB.admins)
```

**Review by utilbird — COMMENTED** [review]
I love how every datum in the interview queue has a `position_in_queue` variable that's decremented instead of just moving their position in the queue list

File: code/modules/mob/dead/new_player/new_player.dm (hunk 6 above)
Comment (ike709) [reply code/modules/mob/dead/new_player/new_player.dm]: /shrug

I didn't write any of this besides redoing the stat panel.

**Comment by Crossedfall** [comment]
Tested on staging and confirmed that it works. There is an interviews tab now and we were able to go through the process. Testing procedure is below. 

## Three bits of feedback from this testing:

1. The extra tab makes the stat-panel even more cluttered than it already is. Integration into the ticketing GUI might be preferable
   1.  The theming/scheme should also probably match the ticketing GUI 
image
2. Pending/unapproved interviewees can still use OOC // This should be disabled for them
3. Denial should softban the user until the bunker is lifted or the round ends

## Testing procedure:

1. Deadmin volunteer (llol)
2. Enable panic bunker with an unachievable living time (9999999)
3. Enable the interview system
4. Have the volunteer connect
5. Confirmed that they cannot see anything extra in the stat panel  
    - image
6. Had them fill out the interview questionnaire 
7. Denied their application
    - Interviewee was allowed to stay connected, they were asked to try again in 3-minutes image
8. Reset and had them try again
9. Approved the interview
10. Confirmed that they were able to join the round

File: code/modules/mob/dead/new_player/new_player.dm [hunk 8]
```diff
@@ -476,3 +484,56 @@
...
+
+	client.tgui_panel.set_verb_infomation(client)
+
+	// Then we create the interview form and show it to the client
+	var/datum/interview/I = GLOB.interviews.interview_for_client(client)
+	if (I)
+		I.ui_interact(src)
+
+	// Add verb for re-opening the interview panel, and re-init the verbs for the stat panel
+	add_verb(src, /mob/dead/new_player/proc/open_interview)
```
Comment (ike709) [inline code/modules/mob/dead/new_player/new_player.dm]:
```suggestion
	add_verb(/mob/dead/new_player/proc/open_interview)
```

**Review by Crossedfall — DISMISSED** [review]
Giving my tentative approval. Passed all the general tests on Staging

File: code/modules/interview/interview.dm [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+	read_only = TRUE
+	GLOB.interviews.approved_ckeys |= owner_ckey
+	GLOB.interviews.close_interview(src)
+	log_admin_private("[key_name(approved_by)] has approved interview #[id] for [owner_ckey][!owner ? "(DC)": ""].")
+	message_admins("<span class='adminnotice'>[key_name(approved_by)] has approved interview #[id] for [owner_ckey][!owner ? "(DC)": ""].</span>")
+	if (owner)
+		SEND_SOUND(owner, sound('sound/effects/adminhelp.ogg'))
+		to_chat(owner, "<font color='red' size='4'><b>-- Interview Update --</b></font>" \
+			+ "\n<span class='adminsay'>Your interview was approved, you will now be reconnected in 5 seconds.</span>")
+		addtimer(CALLBACK(src, .proc/reconnect_owner), 50)
```
Comment (Archanial) [inline code/modules/interview/interview.dm]:
```suggestion
		addtimer(CALLBACK(src, .proc/reconnect_owner), 5 SECONDS)
```

File: code/modules/interview/interview.dm [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+	message_admins("<span class='adminnotice'>[key_name(denied_by)] has denied interview #[id] for [owner_ckey][!owner ? "(DC)": ""]. Retries [can_retry ? "enabled" : "disabled"].</span>")
+	if(can_retry)
+		addtimer(CALLBACK(GLOB.interviews, /datum/interview_manager.proc/release_from_cooldown, owner_ckey), 180 SECONDS)
+		if (owner)
+			SEND_SOUND(owner, sound('sound/effects/adminhelp.ogg'))
+			to_chat(owner, "<font color='red' size='4'><b>-- Interview Update --</b></font>" \
+				+ "\n<span class='adminsay'>Unfortunately your interview was denied. Please try submitting another questionnaire." \
+				+ " You may do this in three minutes.</span>")
+	else
+		addtimer(CALLBACK(GLOB.interviews, /datum/interview_manager.proc/give_the_boot, owner_ckey), 60 SECONDS)
```
Comment (Archanial) [inline code/modules/interview/interview.dm]: imo this should be shorter, like 30 seconds, no need to keep him on the server that long

File: code/modules/interview/interview.dm [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+	winset(owner, null, "command=.reconnect")
+
+/**
+  * Verb for opening the existing interview, or if relevant creating a new interview if possible.
+  */
+/mob/dead/new_player/proc/open_interview()
+	set name = "Open Interview"
+	set category = "Interview"
+	var/mob/dead/new_player/M = usr
+	if (M?.client?.interviewee)
```
Comment (Archanial) [inline code/modules/interview/interview.dm]:
```suggestion
	if (M.client?.interviewee)
```
Can M be ever null?

File: code/modules/interview/interview.dm [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+				GLOB.interviews.enqueue(src)
+				. = TRUE
+		if ("approve")
+			if (usr.client?.holder && status == INTERVIEW_PENDING)
+				src.approve(usr)
+				. = TRUE
+		if ("deny")
+			if (usr.client?.holder && status == INTERVIEW_PENDING)
+				src.deny(usr)
+				. = TRUE
```
Comment (Archanial) [inline code/modules/interview/interview.dm]:
```suggestion
				return = TRUE
```
no need to set . to true if you are not going to do anything else in the proc

File: code/modules/interview/interview_manager.dm [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+/**
+  * Removes a ckey from the server, used when an interview is denied and retries are disabled.
+  *
+  * Arguments:
+  * * ckey - The ckey to remove from the cooldown list
+  */
+/datum/interview_manager/proc/give_the_boot(ckey)
+	for(var/client/C as() in GLOB.clients)
+		if(C?.ckey == ckey)
+			qdel(C)
```
Comment (Archanial) [inline code/modules/interview/interview_manager.dm:196]:
```suggestion
			qdel(C)
			return
```

File: code/modules/interview/interview.dm (hunk 12 above)
Comment (ike709) [reply code/modules/interview/interview.dm]:
```suggestion
				return TRUE
```

File: code/modules/interview/interview.dm [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+			to_chat(owner, "<font color='red' size='4'><b>-- Interview Update --</b></font>" \
+				+ "\n<span class='adminsay'>Unfortunately your interview was denied. Please try submitting another questionnaire." \
+				+ " You may do this in three minutes.</span>")
+	else
+		addtimer(CALLBACK(GLOB.interviews, /datum/interview_manager.proc/give_the_boot, owner_ckey), 60 SECONDS)
+		if (owner)
+			SEND_SOUND(owner, sound('sound/effects/adminhelp.ogg'))
+			to_chat(owner, "<font color='red' size='4'><b>-- Interview Update --</b></font>" \
+				+ "\n<span class='adminsay'>Unfortunately your interview was denied. You will be removed for the round's duration." \
+				+ " You will be kicked in 60 seconds.</span>")
```
Comment (ike709) [inline code/modules/interview/interview.dm]:
```suggestion
		addtimer(CALLBACK(GLOB.interviews, /datum/interview_manager.proc/give_the_boot, owner_ckey), 30 SECONDS)
		if (owner)
			SEND_SOUND(owner, sound('sound/effects/adminhelp.ogg'))
			to_chat(owner, "<font color='red' size='4'><b>-- Interview Update --</b></font>" \
				+ "\n<span class='adminsay'>Unfortunately your interview was denied. You will be removed for the round's duration." \
				+ " You will be kicked in 30 seconds.</span>")
```

File: code/modules/interview/interview.dm [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+				. = TRUE
+		if ("submit")
+			if (!read_only)
+				read_only = TRUE
+				GLOB.interviews.enqueue(src)
+				. = TRUE
+		if ("approve")
+			if (usr.client?.holder && status == INTERVIEW_PENDING)
+				src.approve(usr)
+				. = TRUE
```
Comment (Archanial) [inline code/modules/interview/interview.dm]:
```suggestion
				return TRUE
		if ("submit")
			if (!read_only)
				read_only = TRUE
				GLOB.interviews.enqueue(src)
				return TRUE
		if ("approve")
			if (usr.client?.holder && status == INTERVIEW_PENDING)
				src.approve(usr)
				return TRUE
```

**Review by Archanial — APPROVED** [review]

**Pull request merged**